synapse/handlers/federation.py

```diff
@@ -1089,10 +1170,12 @@ async def _maybe_backfill_inner(
         #   state *before* the event, ignoring the special casing certain event
         #   types have.
 
-        forward_events = await self.store.get_successor_events(list(extremities))
+        forward_event_ids = await self.store.get_successor_events(
+            list(oldest_events_with_depth)
+        )
```

Comment:
Why do we get successor events here?

This seems to be explained by the comment:

> Given we don't have the extremity events themselves, we need to actually check the events that reference them.

Butttt..... we are already getting events that have `depth`. That's the whole point of `get_oldest_events_with_depth_in_room`.

This would only make sense if we returned the backward extremities.

This visibility filtering was added in https://github.com/matrix-org/synapse/commit/71b625d80806886794c5e72f7ff11432e99b736c but the code was still `get_oldest_events_with_depth_in_room` back then.

---

I feel like this is a bug and we don't need to get the forward events first.

Reply:
Changing this code does make some of my federation backfill Complement tests fail which is making it seem like it serves a purpose. But it's not clear to me 😕

Reply:
`get_oldest_events_with_depth_in_room` I *think* returns the event ID of the extremity plus the max depth of all events that reference it? https://github.com/matrix-org/synapse/blob/c167e09fe58d3a256fb1c763b391ad6633d2507d/synapse/storage/databases/main/event_federation.py#L677-L683

Reply:
Can you check `get_oldest_event_ids_with_depth_in_room` in this PR? It's the same function but with comments and is renamed to show that it returns `event_ids` (even though also includes `depth`)

I feel like it returns the closest events to the extremeties because the extremeties aren't persisted in our DB (and have no depth to reference).

Reply:
The query is returning the column `event_backward_extremities.event_id` (in the query the table is aliased as `b`), where the events listed in  `event_backward_extremities` have *not* been persisted yet. There is a bunch of joining to figure out what events point to those and what their max depth is, which is a bit confusing, but doesn't affect the returned event IDa

Reply:
Ahhh, thanks for this! I assumed the `depth` would be associated with the event chosen to return but makes sense that it wouldn't if it was `b.event_id`(`event_backward_extremities.event_id`) being returned.

I've updated the comments to point out that it's an approximate depth based the persisted events connected to the extremity 👍